Mathlib/RingTheory/Valuation/ValExtension.lean

```diff
@@ -0,0 +1,181 @@
+
+theorem val_map_lt_one_iff (x : R) : vA (algebraMap R A x) < 1 ↔ vR x < 1 :=
+  (IsEquiv.val_lt_one val_isEquiv_comap).symm
+
+theorem val_map_eq_one_iff (x : R) : vA (algebraMap R A x) = 1 ↔ vR x = 1 :=
+  (IsEquiv.val_eq_one val_isEquiv_comap).symm
+
+instance id : IsValExtension vR vR where
+  val_isEquiv_comap := by
+    simp only [Algebra.id.map_eq_id, comap_id]
+    rfl
+
+end CoeLemma
+
+section mk'
+
+/--
+When `K` is a field, if the preimage of the valuation integers of `A` equals to the valuation
+integers of `K`, then the valuation on `A` is an extension of valuation on `K`.
+-/
+theorem ofIntegerComap {K A ΓK ΓA : Type*} [Field K] [Ring A]
+    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
+  [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA)
+    (h : vA.integer.comap (algebraMap K A) = vK.integer) : IsValExtension vK vA where
+  val_isEquiv_comap := by
+    rw [isEquiv_iff_val_le_one]
+    intro x
+    rw [← Valuation.mem_integer_iff, ← Valuation.mem_integer_iff]
+    rw [← h]
+    rfl
+
+end mk'
+
+section integer
+
+variable {R A ΓR ΓA : Type*} [CommRing R] [Ring A]
+    [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓA]
+    [Algebra R A] {vR : Valuation R ΓR} {vA : Valuation A ΓA} [IsValExtension vR vA]
+
+instance integerAlgebra : Algebra vR.integer vA.integer where
+    smul r a := {
+      val := r • a,
+      property := by
+        rw [Valuation.mem_integer_iff,
+          show r • ↑a = algebraMap R A r * a by exact (Algebra.smul_def r (a : A))]
+        norm_num
+        apply mul_le_one'
+        · rw [val_map_le_one_iff vR]
+          exact r.2
+        · exact a.2
+    }
+    toFun r := {
+      val := algebraMap R A r,
+      property := by
+        simp only [Valuation.mem_integer_iff]
+        rw [val_map_le_one_iff vR]
+        exact r.2
+    }
+    map_one' := by
+      ext
+      simp
+    map_mul' _ _ := by
+      ext
+      simp
+    map_zero' := by
+      ext
+      simp
+    map_add' _ _ := by
+      ext
+      simp
+    commutes' _ _ := by
+      ext
+      exact Algebra.commutes _ _
+    smul_def' _ _ := by
+      ext
+      exact Algebra.smul_def _ _
+
+@[simp, norm_cast]
+theorem coe_algebraMap_integer (r : vR.integer) :
+    ((algebraMap vR.integer vA.integer) r : A) = (algebraMap R A) (r : R) := by
+  rfl
+
+instance instIsScalarTowerInteger : IsScalarTower vR.integer vA.integer A where
+  smul_assoc x y z := by
+    simp only [Algebra.smul_def]
+    exact mul_assoc _ _ _
+
+theorem integerAlgebra_injective {K A ΓK ΓA : Type*} [Field K] [Ring A] [Nontrivial A]
+    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
+    [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA) [IsValExtension vK vA] :
+    Function.Injective (algebraMap vK.integer vA.integer) := by
+  intro x y h
+  simp only [Subtype.ext_iff, coe_algebraMap_integer] at h
+  ext
+  apply RingHom.injective (algebraMap K A) h
```

Comment:
If you state it like this, it can be an instance (and it is more general)
```lean
instance {K A ΓK ΓA : Type*} [CommRing K] [Ring A]
    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
    [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA) [IsValExtension vK vA]
    [NoZeroSMulDivisors K A] :
    NoZeroSMulDivisors vK.integer vA.integer := by
  refine ⟨fun {x y} e ↦ ?_⟩
  have : (x : K) • (y : A) = 0 := by simpa [Subtype.ext_iff, Algebra.smul_def] using e
  simpa only [Subtype.ext_iff, smul_eq_zero] using this
```
(you can still provide `integerAlgebra_injective` below; and it should be called `algebraMap_injective`.)

Reply:
Thank you. I'll add the instance and rename `integerAlgebra_injective` to `algebraMap_integer_injective`

Reply:
Is there a danger that this theorem name without containing `integer` would make people think of this theorem? (Maybe this theorem will not get into mathlib anyway ...) 
`theorem IsValExtension.algebraMap_injective {K A ΓK ΓA : Type*} [Field K] [Ring A]
    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
    [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA) [IsValExtension vK vA] [Nontrivial ΓA]
    Function.Injective K A`

And this remind me that I should add a theorem that states `Nontrivial ΓA` implies `Nontrivial A`

Reply:
My reason is exactly because this lemma shouldn't be in mathlib. You should get the `Nontrivial A` instance and use `RingHom.injective`.

Also you should generalize `RingHom.domain_nontrivial` to `MonoidWithZero.domain_nontrivial` if you need the last result that you mention.